physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,710 @@
+		serverIDs = append(serverIDs, string(server.ID))
+	}
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	ret := make(map[raft.ServerID]*autopilot.Server)
+	for id, state := range d.followerStates.followers {
+		// If the server is not in raft configuration, even if we received a follower
+		// heartbeat, it shouldn't be a known server for autopilot.
+		if !strutil.StrListContains(serverIDs, id) {
+			continue
+		}
+
+		server := &autopilot.Server{
+			ID:          raft.ServerID(id),
+			Name:        id,
+			RaftVersion: raft.ProtocolVersionMax,
+			Ext:         d.autopilotServerExt(state.NonVoter),
+		}
+
+		switch state.IsDead {
+		case true:
+			d.logger.Debug("informing autopilot that the node left", "id", id)
+			server.NodeStatus = autopilot.NodeLeft
+		default:
+			server.NodeStatus = autopilot.NodeAlive
+		}
+
+		ret[raft.ServerID(id)] = server
+	}
+
+	// Add the leader
+	ret[raft.ServerID(d.localID)] = &autopilot.Server{
+		ID:          raft.ServerID(d.localID),
+		Name:        d.localID,
+		RaftVersion: raft.ProtocolVersionMax,
+		NodeStatus:  autopilot.NodeAlive,
+		Ext:         d.autopilotServerExt(false),
+	}
+
+	return ret
+}
+
+// RemoveFailedServer is called by the autopilot library when it desires a node
+// to be removed from the raft configuration. This function removes the node
+// from the raft cluster and stops tracking its information in follower states.
+// This function needs to return quickly. Hence removal is performed in a
+// goroutine.
+func (d *Delegate) RemoveFailedServer(server *autopilot.Server) {
+	go func() {
+		d.logger.Info("removing dead server from raft configuration", "id", server.ID)
+		if future := d.raft.RemoveServer(server.ID, 0, 0); future.Error() != nil {
+			d.logger.Error("failed to remove server", "server_id", server.ID, "server_address", server.Address, "server_name", server.Name, "error", future.Error())
+		}
+		d.followerStates.Delete(string(server.ID))
+	}()
+}
+
+// SetFollowerStates sets the followerStates field in the backend to track peers
+// in the raft cluster.
+func (b *RaftBackend) SetFollowerStates(states *FollowerStates) {
+	b.l.Lock()
+	b.followerStates = states
+	b.l.Unlock()
+}
+
+// SetAutopilotConfig updates the autopilot configuration in the backend.
+func (b *RaftBackend) SetAutopilotConfig(config *AutopilotConfig) {
+	b.l.Lock()
+	b.logger.Info("updating autopilot configuration", "config", config)
+	b.autopilotConfig = config
+	b.l.Unlock()
+}
+
+// AutopilotConfig returns the autopilot configuration in the backend.
+func (b *RaftBackend) AutopilotConfig() *AutopilotConfig {
+	b.l.RLock()
+	conf := b.autopilotConfig
+	b.l.RUnlock()
+	return conf.Clone()
+}
+
+func (b *RaftBackend) defaultAutopilotConfig() *AutopilotConfig {
+	return &AutopilotConfig{
+		CleanupDeadServers:          false,
+		LastContactThreshold:        10 * time.Second,
+		LastContactFailureThreshold: 24 * time.Hour,
```

Comment:
We might need consensus on the name and value of this.

Reply:
LeftServerLastContactThreshold?

Reply:
I like `LeftServerLastContactThreshold` better than `LastContactFailureThreshold`. I'll make this change.

Reply:
Is this an existing Consul setting? If it's only used for dead server cleanup should we call it `DeadServerLastContactThreshold`?

Reply:
Done.

Reply:
Nope. This isn't an existing Consul setting. Consul doesn't need it since Serf does the dead server detection. For Vault we need to have our own way to flag a node as dead, and this is that. I like `DeadServerLastContactThreshold` over `LeftServerLastContactThreshold`. I'll do that.

Reply:
Done.